### ISSUE TITLE
Does the partial_fit method reset the parameters?

### ISSUE THREAD

**Issue by EliasVansteenkiste** [dnouri/nolearn#272]
I first trained and validated a network offline and saved the parameters of the network to a file. This works quite well.

My goal is now to train my network online starting from the trained network with new information I can only get at runtime. I think the partial_fit method would work well in this situation.

So after defining the model, I read in the parameters from a file:
```
with open(args.model_filename) as f:
	model_norm = pickle.load(f)
netSpec.load_params_from(model_norm['model'])
```

Online, I want to invoke the partial_fit method several times to adapt the network slightly to a changing situation:
`netSpec.partial_fit(train_x,online_correct_y)`
but, I am in doubt about what the partial_fit method does with the weights I loaded from the file. Does it erase the weights before it starts the training?

Setting the learning rate very low, should allow me to get about the same validation scores as I got offline. I tried this, but my validation scores are really low.

So my questions are:

1. Does partial_fit reset the parameters the first time it is invoked?
2. How can I prevent the reset? Or is there another way to achieve the same thing?

Thanks in advance,

Elias

**Comment by BenjaminBossan** [comment]
Neither `partial_fit`, nor `fit`, resets the learned weights (`partial_fit` is actually a `fit` call with only one epoch, and neither -- unlike sklearn's `fit` -- resets the weights).

I assume that you verified that your `netSpec` model performs equally well on the training data as your saved model. In that case, I would consider 2 possibilities first:

* The results are just the way they are. Did you check whether your pickled model's predictions are any better on the new data?
* Since the parameters of the updater are not pickled, it could be that they lead to too much learning (do you use momentum or learning rate decay?).

**Comment by EliasVansteenkiste** [comment]
During training I used adam with an initial learning rate of 0.001.
If I evaluate the pickled model without any modification we get an error rate of 12.45%.

```
learning_rate = theano.shared(np.float32(0.001))
net = NeuralNet(
        layer,
        update=adam,
        update_learning_rate=learning_rate,
        regression=False,
        max_epochs=100,
        verbose=1,
        on_epoch_finished=[EarlyStopping(patience=5),],
)
```

During the online phase I also used adam, but I lowered the learning rate to 0.000001.
The accuracy then drops and we get an error rate of 28.54%.
Should I further lower the learning rate? If I set the learning rate really low than I should be able to get the same accuracy as the pickled model. Is that a correct hypothesis?

However I am in doubt about which update mechanism I should use. Do you have any hints about that?
I think these momentum updating mechanisms presume a "fixed cost function landscape", which is not the case for my application. We could see some strange effects with the acceleration and momentum. It reminds me of the parachute ball game.

**Comment by BenjaminBossan** [comment]
First I have to correct myself, when you pickle the model (in contrast to using `save_params_to`), the updater should also be pickled. Therefore, there should not be anything wrong there.

Two ideas come to my mind next:

1. Set the learning rate to 0. In theory, that should lead to exactly the same error rate as for your initial model. If this is not the case, we know better where to look next.
2. I'm not sure what `EarlyStopping` code you use, but that could potentially interfere with training. Could you test whether the effect still occurs without `EarlyStopping`?

**Comment by EliasVansteenkiste** [comment]
When the network is trained we used the following update mechanism:  
```
net = NeuralNet(
        layer,
        update=adam,
        update_learning_rate=0.001,
        #update_momentum=0.9,
        regression=False,
        max_epochs=100,
        verbose=1,
        on_epoch_finished=[EarlyStopping(patience=5),],)

class EarlyStopping(object):
    def __init__(self, patience=100):
        self.patience = patience
        self.bestValid = np.inf
        self.bestValidEpoch = 0
        self.bestWeights = None

    def __call__(self, nn, trainHistory):
        currentValid = trainHistory[-1]['valid_loss']
        currentEpoch = trainHistory[-1]['epoch']
        if currentValid < self.bestValid:
            self.bestValid = currentValid
            self.bestValidEpoch = currentEpoch
            self.bestWeights = nn.get_all_params_values()
        elif self.bestValidEpoch + self.patience < currentEpoch:
            print("Early stopping.")
            print("Best valid loss was {:.6f} at epoch {}.".format(
                self.bestValid, self.bestValidEpoch))
            nn.load_params_from(self.bestWeights)
            raise StopIteration()
```

During online training we use this update mechanism:
```
net = NeuralNet(
        layer,
        update=adam,
        update_learning_rate=0,
        regression=False,
        max_epochs=100,
        verbose=1,)
```
and we use the same layer definitions of course. I don't think the early stopping is a problem as we don't use it during the online training.

The offline trained network achieves an error rate of 12.43%.
After training with a learning rate of 0, we get an error rate of 13.89%.
So we closed a big gap and so clearly something goes wrong during online training. Unfortunately I can't say yet what is going wrong as I am still investigating that, but why do we still have a gap of 1.46% between the offline trained network and the online trained network if the learning rate is 0?

**Comment by BenjaminBossan** [comment]
So do I get it right that your new net has the same error rate as the initial one directly after loading but becomes worse over time despite a learning rate of 0? If that were so, I would be quite surprised. I would need to have a closer look at the adam code but my understanding is that there should be no change with a learning rate of 0.

Maybe you could try a different, simple updater such as pure SGD. Does the performance still decrease? Even with an error rate of 0?

**Comment by dnouri** [comment]
Op hasn't responded in a while.  Pleae reopen if coming back with an answer.

**Comment by smar10** [comment]
@EliasVansteenkiste do you discover the problem?
